pkg/ccl/sqlproxyccl/forwarder.go

```diff
@@ -105,3 +129,78 @@ func (f *forwarder) Close() {
 	// has already been closed.
 	f.serverConn.Close()
 }
+
+// handleClientToServer handles the communication from the client to the server.
+// This returns a context cancellation error whenever the forwarder's context
+// is cancelled, or whenever forwarding fails. When ForwardMsg gets blocked on
+// Read, we will unblock that through our custom readTimeoutConn wrapper, which
+// gets triggered when context is cancelled.
+func (f *forwarder) handleClientToServer() error {
+	for f.ctx.Err() == nil {
+		if _, err := f.clientInterceptor.ForwardMsg(f.serverConn); err != nil {
+			return err
+		}
+	}
+	return f.ctx.Err()
+}
+
+// wrapClientToServerError overrides client to server errors for external
+// consumption.
+//
+// TODO(jaylim-crl): We don't send any of these to the client today,
+// unfortunately. At the moment, this is only used for metrics. See TODO in
+// proxy_handler about sending safely to avoid corrupted packets. Handle these
+// errors in a friendly manner.
+func (f *forwarder) wrapClientToServerError(err error) error {
```

Comment:
nit: wrapClientToServerError and wrapServerToClientError can be implemented as functions instead of methods.